### DISCUSSION TITLE
Help: how to use rewrite()

### DISCUSSION THREAD

**Discussion by vforchi** [nom-tam-fits/nom-tam-fits#714] [Q&A]
Hi,

I am trying to modify a FITS header by adding a few keywords, and I would like to minimize the disk I/O. Is this correct?
```
        var fits = new Fits("some/location");
        var hdu = fits.getHDU(0);
        hdu.header.addValue("SOMEKWD", "value", "comment");
        hdu.setChecksum();
        if (hdu.rewriteable()) {
            hdu.rewrite();
        } else {
            fits.write(file.url);
        }
        fits.close();
```
I used `rewritable()` because I noticed that `hdu.rewrite()` can throw an Exception if it needs to write a new block (I assume), but I'm not sure this is correct.

**Comment by attipaci** [comment]
Hi @vforchi,

Yes, that is correct. And, your example code looks alright to me.

The purpose of `rewritable()` is exactly to check if the data can be re-written in place (in the same FITS storage area). If you look at its Javadoc it says:

> Checks if we can write this element back to its source. An element can
> only be written back if it is associated to a random accessible input and
> the current size fits within the old block size.

If `rewritable()` returns `true` then `rewrite()` is not expected to throw an exception (but it still can(!), for example it the file system gets unmounted, or is read-only, or full -- or if the FITS is corrupted).

Of course, you don't really need `rewritable()` per se as you can do a `try`/`catch` block too mostly to the same effect:

```java
  try {
    hdu.rewrite();
  } catch (FitsException e) {
    // the HDU could not be re-written either because it was read not from a random accessible input
    // or because it has outgrown the size available in the FITS, -OR- because its integrity has been 
    // compromised.
    // You can either delete the original and try write anew to the same path, or you can try write to
    // a different path (leaving the original intact).
    ...
  } catch (IOException e) {
    // It should have worked, but some non FITS related I/O issue prevented it from happening.
    ...
  }
```

The two are generally equivalent, since `rewrite()` first calls `rewritable()` internally anyway to check, and will  throw a `FitsException` if it returned `false`. Ultimately the only difference  between the two approaches is that `rewrite()` may throw a `FitsException` for other reasons also, such as problems with the FITS' integrity. So it depends if you want to handle integrity issues separately or together with the logistical disability to write back to the same file... 

Does this help?

**Comment by attipaci** [comment]
I'll convert this to a discussion if you don't mind.

**Comment by vforchi** [comment]
Thanks. For completeness:
```
  try {
    hdu.rewrite();
  } catch (FitsException e) {
    fits.write(file.url);
  } catch (IOException e) {
    // nothing to do but to rethrow, I guess
  } finally {
    fits.close();
  }
```
Correct?

**Reply by attipaci** [reply]
Yes, that looks like perfection to me. :-) Sorry for the slow response. My attention has been elsewhere.
